### PR TITLE
Figure out when we underestimate the heap page size

### PR DESCRIPTION
**Pull request by mppf**
Addresses a problem with the muxed tasking layer when using GASNet on a Cray XC system but without a hugepages module loaded.

In that configuration even Hello World would crash.  The problem was that in that configuration, GASNet arranges for the heap to be on huge pages.  Then muxed tries to create guard pages, but the mprotect call to do so creates a guard hugepage (much too large) instead of a guard page. The result is a core dump when the following code tries to use adjacent memory.

This patch adds logic to chplsys.c to try calling mprotect and accessing memory in sequence in a way that experimentally determines the page size of the Chapel heap. This new logic is run from chpl_mem_init.

Verified a simple Chapel program running getHeapPageSize() returns the correct value in these configurations:
 * a linux 4.4 system (using a custom comm-none that allocates huge pages).
 * Mac OS X 10.10.5 (using a custom comm-none that allocates huge pages and using SIGBUS instead of SIGSEGV)
 * a Cray XC with comm=gasnet tasks=fifo and with CHPL_RT_CALL_STACK_SIZE=128k or 2MB (this configuration did not work before)
 * a Cray XC with comm=gasnet tasks=muxed without a hugepages module loaded
 * a Cray XC with comm=ugni tasks=muxed with or without a hugepages module loaded

Verified that Hello World works in these configurations:
 * a Cray XC with GASNet or UGNI and tasks=muxed with or without a hugepages module
 * a 64-bit Cygwin system
 * a FreeBSD 10 system
 * an omnios (solaris clone) system

Passed full local testing.

Reviewed by @gbtitus - thanks!

### CODE REVIEW

File: runtime/src/chplsys.c [hunk 1]
```diff
@@ -73,11 +76,164 @@ size_t chpl_getSysPageSize(void) {
...
+  // install signal handler
+  {
+    struct sigaction act = { .sa_sigaction = check_page_size_segv_handler,
+                             .sa_flags = SA_SIGINFO };
+    rc = sigaction(SIGSEGV, &act, &check_page_size_oldact);
+    if (rc!=0)
+      return -1;
+  }
+
+  // Set the pointer to 0 to avoid false positives:
```
Comment (gbtitus) [inline runtime/src/chplsys.c]: Strictly speaking, it's the pointee we're setting to 0 here, not the pointer.

File: runtime/src/chplsys.c [hunk 2]
```diff
@@ -52,6 +52,9 @@
 #include <errno.h>
 #include <unistd.h>
 
+#include <signal.h>
```
Comment (gbtitus) [inline runtime/src/chplsys.c]: Generally the include files are in two clumps, the local ("") ones and then the system (<>) ones.  Then within the clumps the files are in alphabetical order.  This makes everything easier to find.  It's an unending battle trying to keep things in this order (i noted that even now there are some mis-orderings at least within the system ones), but let's keep trying.  So, would you please move these three new ones appropriately?

**Comment by gbtitus** [comment]
Looks really good --  I just had a couple of line comments.  It's good to go, with those addressed.

File: runtime/src/chplsys.c (hunk 2 above)
Comment (mppf) [inline runtime/src/chplsys.c]: I'm not sure keeping them in order is a worthwhile trade-off for adding more #ifdef __linux__. I think I would rather see all platform-specific headers included first and then all others. We could still have a group for platform-specific things...

**Comment by mppf** [comment]
@gbtitus - I just noticed that FIFO+Gasnet also does not work e.g. with CHPL_RT_CALL_STACK_SIZE=2M - I think our default 8M stacks have been hiding this issue.

File: runtime/src/chplsys.c (hunk 2 above)
Comment (gbtitus) [inline runtime/src/chplsys.c]: I'm not seeing the need for more `#ifdef` linux constructs.  None of these are in such a construct now.  Wouldn't `#include <signal.h>` go directly before `#include <stdio.h>`, `#include <sys/mman.h>` go directly after `#include <string.h>`, and `#include "chpl-align.h"` go before `#include "chpl-mem.h"`?

Comment (mppf) [inline runtime/src/chplsys.c]: have a look at the version I just merged, which puts them in an ordering I can be happy about too.

The reason I did that is that 'signal' sorts between 'pthread' and 'sched'. But maybe you have some partial ordering in mind...

**Comment by gbtitus** [comment]
I'm fuzzy on "FIFO+Gasnet also does not work ...".  Doesn't work on Cray X* with the GASNet aries conduit?  And does this PR fix that failure, or is there more to do?  (A side question, more for curiosity's sake than anything else: do you know how GASNet chooses a hugepage size to use with the aries conduit and/or what hugepage size it ends up with?)

**Comment by gbtitus** [comment]
I'm not happy with the `#include` reordering. The runtime HAS a `#include` ordering already: a block of local (`""`) includes, then a block of system (`<>`) includes, with the files within each block in alphabetical order.  While adherence to this ordering is certainly far from perfect in the sources, let's do our best to stick to it, because similarity across the source files makes maintenance easier.

**Comment by noakesmichael** [comment]
I am with Greg on this.  This is the convention that’s been in place at
every company I have worked at with the clarification that the first
#include in, for example, foo.cpp should be ‘#include “foo.h”’ followed
by the remaining locals.

Putting foo.h first in foo.cpp ensures that foo.h doesn’t accidentally
depend on some file that it doesn’t include for itself.


I am never quite sure what to do about sorting files when some
are capitalized and others are not i.e. whether to sort ignoring
case or not.

This shouldn’t be a runtime vs. compiler issue.

Mike



On Dec 2, 2016, at 7:47 AM, Greg Titus <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm not happy with the #include reordering. The runtime HAS a #include ordering already: a block of local ("") includes, then a block of system (<>) includes, with the files within each block in alphabetical order. While adherence to this ordering is certainly far from perfect in the sources, let's do our best to stick to it, because similarity across the source files makes maintenance easier.

**Comment by mppf** [comment]
@gbtitus, @noakesmichael - we are not free to re-order the #includes in the runtime in an abritrary order. The reason is that there are #defines that affect the later #includes. The #define _GNU_SOURCE is one such #define in chplsys.c but there are others in sys_basic.h. Since chplrt.h includes sys_basic.h, we normally include chplrt.h first.

After that I have no problem including the corresponding header file for the .c file in question.

The issue we should be wrestling with here is whether it is better to repeat a #ifdef __linux__ multiple times in order to keep system includes in sorted order - or if it is better to put the ifdef __linux__ headers in a single block. It sounds like @gbtitus would prefer several ifdef blocks, which is fine. Is that the case, Greg? I just need a clear answer to that question.

**Comment by gbtitus** [comment]
First, as to the meta-question: yes, I'd prefer one ordered list of system include files, and if that requires multiple `#ifdef` blocks then so be it.

Second, specific for this PR, in my review feedback I wasn't asking that you reorder all the existing `#includes` to follow the rule I'd described.  You can if you want to, as a "while I was here" change, but it's not necessary.  I was only asking that you slot in your new `#includes` as best you could so that things didn't get worse.

**Comment by mppf** [comment]
Re FIFO+GASNet, yes I was talking about a Cray XC using the aries conduit. And this PR resolves the issue.

**Comment by mppf** [comment]
This PR is ready to merge but I'm waiting to merge it until Monday.

**Pull request merged**